docs/design/enhance-pluggable-policy.md

```diff
@@ -0,0 +1,175 @@
+# Enhance Pluggable Policy
+While trying to develop an authorization plugin for Kubernetes, we found a few places where API extensions would ease development and add power.  There are a few goals:
+ 1.  Provide an authorization plugin that can evaluate a .Authorize() call based on the full content of the request to RESTStorage.  This includes information like the full verb, the content of creates and updates, and the names of resources being acted upon.
+ 1.  Provide a mechanism for a user to determine whether or not he can take an action without attempting the action itself.
+ 1.  Provide a way for disconnected pieces of the system to ask whether a user is permitted to take an action without running in process with the API Authorizer.  For instance, a proxy for exec calls could ask whether a user can run the exec they are requesting.
+ 1.  Provide a mechanism to discover who can perform a given action on a given resource.  This is useful for answering questions like, "who can create replication controllers in my namespace".
+
+This proposal adds to and extends existing API to so that authorizers may provide the functionality described above.  It does not attempt to describe how the policies themselves can be expressed, that is up the authorization plugins themselves.
+
+
+## Enhancements to existing Authorization interfaces
+The existing Authorization interfaces are described here: [authorization.md](../authorization.md).  A couple additions will allow the development of an Authorizer that matches based on different rules than the existing implementation.
+
+### Request Attributes
+The existing authorizer.Attributes only has 5 attributes (user, groups, isReadOnly, kind, and namespace).  If we add more detailed verbs, content, and resource names, then Authorizer plugins will have the same level of information available to RESTStorage components in order to express more detailed policy.  The replacement excerpt is below.
+
+A request has 7 attributes that can be considered for authorization:
```

Comment:
Have we had any serious disagreement to this proposal so far?  I haven't
heard any....

On Sep 4, 2015, at 10:49 AM, David Eads <notifications@github.com> wrote:

In docs/design/enhance-pluggable-policy.md
<https://github.com/kubernetes/kubernetes/pull/12209#discussion_r38758883>:

> +While trying to develop an authorization plugin for Kubernetes, we found a few places where API extensions would ease development and add power.  There are a few goals:
> + 1.  Provide an authorization plugin that can evaluate a .Authorize() call based on the full content of the request to RESTStorage.  This includes information like the full verb, the content of creates and updates, and the names of resources being acted upon.
> + 1.  Provide a mechanism for a user to determine whether or not he can take an action without attempting the action itself.
> + 1.  Provide a way for disconnected pieces of the system to ask whether a user is permitted to take an action without running in process with the API Authorizer.  For instance, a proxy for exec calls could ask whether a user can run the exec they are requesting.
> + 1.  Provide a mechanism to discover who can perform a given action on a given resource.  This is useful for answering questions like, "who can create replication controllers in my namespace".
> +
> +This proposal adds to and extends existing API to so that authorizers may provide the functionality described above.  It does not attempt to describe how the policies themselves can be expressed, that is up the authorization plugins themselves.
> +
> +
> +## Enhancements to existing Authorization interfaces
> +The existing Authorization interfaces are described here: [authorization.md](../authorization.md).  A couple additions will allow the development of an Authorizer that matches based on different rules than the existing implementation.
> +
> +### Request Attributes
> +The existing authorizer.Attributes only has 5 attributes (user, groups, isReadOnly, kind, and namespace).  If we add more detailed verbs, content, and resource names, then Authorizer plugins will have the same level of information available to RESTStorage components in order to express more detailed policy.  The replacement excerpt is below.
> +
> +A request has 7 attributes that can be considered for authorization:

note to self, add nonResourceURL similar to openshift to still allow full
filtering of the entire mux.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12209/files#r38758883>.